tools/file_packager.py

```diff
@@ -952,54 +952,73 @@ def generate_js(data_target, data_files, metadata):
-              xhr.addedTotal = true;
-              if (!Module['dataFileDownloads']) Module['dataFileDownloads'] = {};
-              Module['dataFileDownloads'][url] = {
-                loaded: event.loaded,
-                total: size
-              };
-            } else {
-              Module['dataFileDownloads'][url].loaded = event.loaded;
+        Module.dataFileDownloads ??= {};
+        fetch(packageName)
+          .catch((cause) => Promise.reject(new Error(`Network Error: ${packageName}`, {cause}))) // If fetch fails, rewrite the error to include the failing URL & the cause.
+          .then((response) => {
+
+            let loaded = 0;
+
+            if (!response.ok) {
+              return Promise.reject(new Error(`${response.status}: ${response.url}`));
             }
-            var total = 0;
-            var loaded = 0;
-            var num = 0;
-            for (var download in Module['dataFileDownloads']) {
-            var data = Module['dataFileDownloads'][download];
-              total += data.total;
-              loaded += data.loaded;
-              num++;
+
+            // If we're using the polyfill, readers won't be available...
+            if (!response.body && response.arrayBuffer) {
+              response.arrayBuffer().then(callback);
+              return;
             }
-            total = Math.ceil(total * Module['expectedDataFileDownloads']/num);
-            Module['setStatus']?.(`Downloading data... (${loaded}/${total})`);
-          } else if (!Module['dataFileDownloads']) {
-            Module['setStatus']?.('Downloading data...');
-          }
-        };
-        xhr.onerror = (event) => {
-          throw new Error("NetworkError for: " + packageName);
-        }
-        xhr.onload = (event) => {
-          if (xhr.status == 200 || xhr.status == 304 || xhr.status == 206 || (xhr.status == 0 && xhr.response)) { // file URLs can return 0
-            var packageData = xhr.response;
-            callback(packageData);
-          } else {
-            throw new Error(xhr.statusText + " : " + xhr.responseURL);
-          }
-        };
-        xhr.send(null);
+
+            const reader = response.body.getReader();
+            const headers = response.headers;
+
+            const size = Number(headers.get('Content-Length') ?? packageSize);
+            const chunks = [];
+
+            const iterate = () => reader.read().then(handleChunk).catch((cause) => {
+              return Promise.reject(new Error(`Unexpected error while handling : ${response.url} ${cause}`, {cause}));
+            });
+
+            const handleChunk = ({done, value}) => {
+              if (!done) {
+                chunks.push(value);
+                loaded += value.length;
+                Module.dataFileDownloads[packageName] = Module.dataFileDownloads[packageName] ?? {};
+                Module.dataFileDownloads[packageName].loaded = loaded;
+                Module.dataFileDownloads[packageName].total = size;
```

Comment:
Can't you just do `Module.dataFileDownloads[packageName] = {loaded, total: size}` here?    There are not other fields are there?

Reply:
Also I think `dataFileDownloads` needs to be in quotes here.

Reply:
@sbc100 I'd rather just create the object once and then re-use the fields, instead of re-creating it and garbage-collecting the same two fields over and over again.